Parallel remeshing

fsalmon001 opened the pull request:
[Quick_user_guide.odt](https://github.com/user-attachments/files/23962476/Quick_user_guide.odt)
Integration of parallel remeshing into nextsim using parmmg2d. The parallel interpolation is also added.

The use of BAMG is still possible.

There will be some minor differences with the develop branch even with bamg due to changes in the explicitSolve function (induce only machine precision errors).

To use parmmg2d, you must compile parmmg2d. It should be done by adding in the container:

```
wget https://github.com/MmgTools/parMmg2D/archive/refs/heads/nextsim.tar.gz
tar -xzf nextsim.tar.gz
cd parMmg2D-nextsim
mkdir build
cd build
cmake .. -DCMAKE_INSTALL_PREFIX=/opt/nextsim_gnu/parmmg2d/
make -j
make install -j
```

I will edit this message and add a quick user guide for using the parallel remeshing this afternoon.

fsalmon001 committed:
Parallel remeshing

tdcwilliams committed:
compiling with mmg

tdcwilliams committed:
simplify Makefile after MMG make install fixed; other minor changes

tdcwilliams committed:
better defaults for numerics.metric_coef_min, numerics.metric_coef_max

tdcwilliams committed:
make defaults of numerics.metric_coef_min and numerics.metric_coef_max closer together

fsalmon001 commented:
I tried your case @tdcwilliams, but without the ocean model (constant instead) and the ice-type is piomas. I don't have the drifter file neither so I did not use drifters. And it works.

I think I have only seen this kind of error, and if I remember well, it is related to an issue with a netcdf file (netcdf uses the argument ncuts in some functions).

fsalmon001 commented:
I just thought of it @tdcwilliams, @einola, but when the paper will be under review, maybe the reviewers should have access to a version with MMG and BAMG to check and compare the results? 

Maybe we could let BAMG in this version, and I do another pull request to remove BAMG from the code after? My post-doc ends at the end of January, so we have enough time, it will be a quick PR. 

I also had a look on the interpolation function from BAMG, I think I can code another one, but not sure so far that I will be as quick as the original one. I will keep you informed

fsalmon001 commented:
I go a bit deeper in the bamg code for the interpolation and I think it considers that some nodes are outside the domain, and actually, they are not. The error seems to me unbelievable so tell me if you understand the same thing please @tdcwilliams.

In the function InterpFromMeshToMesh2dx, they retrieve the area of the triangle through the attribute det of the triangle (twice the area). But, this is an integer! So, sometimes, it is between 0 and 1 et thus tb.det = 0. Then, line 107 of the file, we don't have tb.det > 0 (I tested it) and therefore the algorithm just looks for another adjacent triangle in the function CloseBoundaryEdge where you have the error kkk>=1000.

I coded a function to avoid this. It seems to work but I have to test it a bit more before doing a commit, maybe tomorrow. However, I can't have exactly the same results as with the BAMG function because of this issue that I don't have in my function.

fsalmon001 committed:
Replacing BAMG node interpolation with a new internal nextsim function in interpolation.cpp

fsalmon001 committed:
Merge remote-tracking branch 'origin/parallel_remeshing' into parallel_remeshing

fsalmon001 committed:
Correct bug with xDelete from BAMG

fsalmon001 commented:
Hi @tdcwilliams, does my case without modification crash also for you? I don't succeed in making it crash here, also on our supercomputer.

fsalmon001 committed:
Improve interpolation for points outside every triangle

tdcwilliams commented:
Hi @fsalmon001 and happy new year.
I tried your updated code and you seem to have fixed the crash I was getting. I'll test it out a bit more and see how it goes for different meshes and number of cpus.

fsalmon001 committed:
Correct issues with machine epsilon in the metric computation

fsalmon001 commented:
Hi @tdcwilliams, I made a commit to avoid little issues in the metric computation. Due to machine epsilon errors, the metric could sometimes disrespect slightly the constraint on the target number of vertices.

einola commented:
Hi guys.

I've been trying to run a 1 km mesh - a slightly larger one than the one you used in the paper @fsalmon001. Unfortunately, I always get ``Major error in the InterpFromMeshToMesh2d function.``. 

I think we should just put a pin in the new ``InterpFromMeshToMesh2d`` function and go back to the bamg version. Especially since your contract is soon over.

B.t.w. I tried reverting just 4bfb1827ba44f395f61367c046cccf2ec95dcb0b, and also first f175c355a69d7731627a9e09106173125a7e2f39 and then 4bfb1827ba44f395f61367c046cccf2ec95dcb0b. My first attempt failed with a git error. The second revert worked, but the model crashed with a seg-fault.

einola commented:
Ok, so the problem has to do with using ``ice-type=topaz4_rean-osisaf-icesat``. In particular, it seems that the model chockes when it tries to interpolate the ice-type files, i.e. the ``ice_osisaf_type_elements`` in ``dataset.cpp``.

If I choose ``ice-type=constant``, then the model runs just fine.

einola commented:
No worries - it's the weekend anyway :)

I've uploaded the files that I used - including the config file - to ftp.nersc.no/pub/einola/nextsim_dbg, if you want to try it on Monday.

fsalmon001 committed:
Debug loading of coarse data when using a great number of CPUs

tdcwilliams commented:
Hi @fsalmon001 
it got a bit further (about 3 days), and didn't give "Major error" with 64 cpus, but still crashed with segmentation fault.
With 128 cpus it crashed at init time (PIOMAS dataset), also with segmentation fault

fsalmon001 commented:
PIOMAS 1996 with a start date of 2 February and with 64 and 128 processors works here (I only imported february)

einola commented:
It's not clear if it's _exactly_ the same location. It's in the initialisation, though.

I'm investigating, but I have to take a break from this now until after lunch sometime.

tdcwilliams commented:
Hi @fsalmon001 
I guess the forcing doesn't have the same problem as piomas and icesat as it is higher resolution, but maybe it will also crash for a higher number of cpu's, so in my opinion it would be good to solve it.
However, the runtime crash in `compute_optimal_metric` should probably be the main priority. I am indeed using those metric settings that you wrote
```
[numerics]
metric_coef_min = 0.9
metric_coef_max = 1.1
```

einola commented:
> Yes I agree with that @einola, but then, at time step XX, the code reloads again the data to impose the new forcing properties and does a new interpolation isn't it? So I don't understand why it does not crash at this moment if it crashes during the initialisation in the interpolation, except if some data are only read initially. But if it works like that, it's a good point.

Yes, I see. Well, the forcing data (winds, currents, etc) are much higher resolution than the initial conditions we had problems with. So that still works fine. I don't expect we'll ever be in a situation where the forcing data is so coarse that we have this problem with those.

fsalmon001 committed:
Debug buffer-overflow in interpolation functions

tdcwilliams commented:
Hi @fsalmon001, 
with this change it crashes again.

tdcwilliams commented:
didn't crash but gave
`INACCURACIES OR ERROR IN THE METRIC COMPUTATION: 583356 is not equal to 554498`.

NB I used the latest edit of the code

fsalmon001 commented:
Ok, so the solution is not the first k and the issue is that there is no solution. I have run a lot of cases and I never faced this. I don't know how it is possible. 

Basically, this means that the number of vertices will increase a bit in the new mesh. So it is not a huge issue. The next regrid step, do you have this message again?

tdcwilliams commented:
I don't know about the mesh, but the moorings outputs seem ok.
This is still init from piomas.

fsalmon001 commented:
Ok, I will try to understand if this is theoretically possible. At least, this is not a critical error, rather an inaccurracy on the wanted metric.

tdcwilliams committed:
check if partitioned first: otherwise is_ghost is unitialised

fsalmon001 committed:
Debug buffer-overflow in boundary flags

fsalmon001 committed:
Debug the metric when the solution is below the machine precision

fsalmon001 committed:
Correct the cohesion when the mesh is not uniform

fsalmon001 committed:
Cleaner addition of MMG options

tdcwilliams committed:
stop part of header being declared twice (only happens if USE_OASIS=1)

einola committed:
Fixes to conservative remapping scheme

There was a problem with the function finding if points are inside or
outside the polygon. This fixes that, including some workarounds. I need
to document it better, and some improvements are still needed.

einola committed:
Don't calculate remeshing weights on elements that don't move

It's not quite the same as skipping elements that have only undergone
translation, but it helps with the small delta errors I've been having.

einola committed:
Do a better job when we're uncertain if a point is inside or not

Following a previous commit, this tells checkIfInside what to return if
the point is right on the border.

einola committed:
Note Spelling

einola committed:
Improve intersection algorithm

A comparison with zero is not very useful on a finite-precision machine!
Using 1e-6 seems to work well.

einola committed:
Sensible limits for mmgopt->hmin and hmax

These seem to be ints and the compiler complains that 1e15 is too big.
So, I limited the size of elements to 1 m at the smallest and 100
thousand km (1e7) at the largest. We don't want elements that cover the
entire earth!

einola committed:
Merge remote-tracking branch 'origin/develop' into parallel_remeshing

einola committed:
Make the model stop at "Major error ..."

The "major error" in InterpFromMeshToMesh2d should be fatal.

einola committed:
A minor optimisation for ConservativeRemappingWithWeights

I was fiddling with it ans saw that I had already done something
similar in ConservativeRemappingFromMeshToMesh. It probably won't have
much of an effect. Putting it here, since these functions have been
massively rewritten anyway.

einola committed:
Throw in a lot of 'const's in to interpolation.cpp

Because it's a good thing!

einola committed:
Limit the reach of the intersection algorithm

checkIfIntersecting now ignores intersections that are exactly on the
vertex. It never should have included those, and doing it like this is
very marginally better in terms of total error.

einola commented on model/options.cpp:
```diff
@@ -75,13 +79,31 @@ namespace Nextsim
+            ("numerics.metric_coef_min", po::value<double>()->default_value( 0.9999 ),
+                "Scale factor for minimum element size.")
+            ("numerics.metric_coef_max", po::value<double>()->default_value( 1.0001 ),
+                "Scale factor for maximum element size.")
```
Hi @fsalmon001,

I needed to change them, because otherwise the model would crash with a message saying that ``mmgopt->hmin`` and ``mmgopt->hmax`` were the same value (I don't have the exact message now).

But you're right, I didn't get that for the 10 km mesh. I can check at which resolution this message appears, if that helps. Do I understand correctly that the optimal values for these parameters are expected to be resolution-dependent?

fsalmon001 replied:
Could you give me the exact message please when possible?

With values very close to 1, it is not resolution-dependent. With your values, it will be. However, the difference is only the computation of the metric, which should almost be constant so the output mesh should be about the same.

einola replied:
```
  ## Error: hmin value must be strictly lower than hmax one (hmin = 1410.000000  hmax = 1410.000000 ).
```

Repeated very often for a mesh with a resolution of 1.25 km.

fsalmon001 replied:
Ok, the warning comes from MMG, it is not a critical error. However, this should always appear, even with 10-km meshes. For me, this is an inaccuracy inside MMG which should tolerate constant metrics.

einola replied:
Ah, I see. Ok, but then I'll remove this from my config files and try to run the high-resolution case. It's a bit strange that it doesn't appear all the time, as you expect, but let's not worry about that for now.

Thanks for the info!

einola committed:
Bugfix for check if nodes move

Fixes a stupid bug, where I didn't calculate the displacement of the
element correctly.

einola committed:
Fix previous fix

I didn't test it properly before pushing.
